py/objtype.c

```diff
@@ -828,6 +828,19 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
             dest[0] = MP_OBJ_NEW_QSTR(self->name);
             return;
         }
+        if (attr == MP_QSTR___dict__) {
+            mp_obj_t attr_dict = mp_obj_new_dict(0);
+            if (self->locals_dict != NULL) {
+                mp_map_t *map = mp_obj_dict_get_map(MP_OBJ_FROM_PTR(self->locals_dict));
+                for (mp_uint_t i = 0; i < map->alloc; ++i) {
+                    if (MP_MAP_SLOT_IS_FILLED(map, i)) {
+                        mp_obj_dict_store(attr_dict, map->table[i].key, map->table[i].value);
+                    }
+                }
+            }
```

Comment:
The idea was to stay close to #1757 code, beside not copying `locals_dict` means you can modify it only if it is not null, which is pretty strange from the Python point of view (e.g. sometime you can modify `__dict__`, sometime the modification is not taken into account...)
Anyway I've simplify my implementation to follow you're advice. I you prefer the first solution in the end, just tell me and I'll revert ;-)